Реализовал API, в котором эндпоинты соответствовуют вызовам базовых методов интерфейса TaskManager.

ViktorSalk opened the pull request:
1) Создал класс HttpTaskServer — он будет слушать порт 8080 и принимать запросы. 
2) Создал класс обработчик BaseHttpHandler
3) Создал класс DurationAdapter для сериализации и десериализации объектов Duration. 
4) Создал key-value сервер для хранения данных.
5) Создал классы TaskHandler, SubtaskHandler, EpicHandler, для обработки HTTP-запросов для разных типов задач. 
6) Создал классы HistoryHandler, PrioritizedHandler для обработки HTTP-запросов в работе с историей просмотров и приоритезацией задач 
7) Сделал тесты по новым классам для различных сценариев.

Viktor3642 committed:
Реализовал API, в котором эндпоинты соответствовуют вызовам базовых методов интерфейса TaskManager.
1) Создал класс HttpTaskServer — он будет слушать порт $8080$ и принимать запросы.
2) Создал класс обработчик BaseHttpHandler
3) Создал класс DurationAdapter для сериализации и десериализации объектов Duration.
4) Создал key-value сервер для хранения данных.
5) Создал классы TaskHandler, SubtaskHandler, EpicHandler, для обработки HTTP-запросов для разных типов задач.
6) Создал классы HistoryHandler, PrioritizedHandler для обработки HTTP-запросов в работе с историей просмотров и приоритезацией задач
7) Сделал тесты по новым классам для различных сценариев.

afomkina commented on src/main/java/history/HistoryHandler.java:
```diff
@@ -0,0 +1,62 @@
+            String method = exchange.getRequestMethod();
+
+            if (!"GET".equals(method)) {
+                sendResponse(exchange, "{\"error\": \"Method not allowed\"}", 405);
```
Задание со `"*"`:
Можно создать отдельное перечисление для HTTP кодов и использовать его вместо "голых" цифр. 
Так код будет выглядеть более читаемо и понятней. 
Примеры можно посмотреть в главе [Конструкторы, поля и методы перечисления](https://metanit.com/java/tutorial/3.8.php#:~:text=%D0%9A%D0%BE%D0%BD%D1%81%D1%82%D1%80%D1%83%D0%BA%D1%82%D0%BE%D1%80%D1%8B%2C%20%D0%BF%D0%BE%D0%BB%D1%8F%20%D0%B8%20%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D1%8B%20%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%81%D0%BB%D0%B5%D0%BD%D0%B8%D1%8F)

ViktorSalk replied:
Спасибо) 
Создал отдельное перечисление для HTTP-кодов и использовать его вместо «голых» цифр в задачх, подзадачах, эпиках, приоритетах.

afomkina commented on src/main/java/history/HistoryHandler.java:
```diff
@@ -0,0 +1,62 @@
+            sendResponse(exchange, gson.toJson(history), 200); // 200 для GET запроса
+        } catch (Exception e) {
+            System.out.println("Произошла ошибка: " + e.getMessage());
+            e.printStackTrace();
```
Печатать стек трейс обычно изличшне, лучше более точно настроить логирование

ViktorSalk replied:
Убрал лишний стек-трейсинг.

afomkina commented on src/main/java/history/HistoryHandler.java:
```diff
@@ -0,0 +1,62 @@
+
+            List<Task> history = taskManager.getHistory();
+            sendResponse(exchange, gson.toJson(history), 200); // 200 для GET запроса
+        } catch (Exception e) {
```
Нужно переименовать однобуквенные переменные в коде

ViktorSalk replied:
Переименовал однобуквенные переменные в коде.

afomkina commented on src/main/java/history/HistoryHandler.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        try {
```
Лучше использовать [try with resources](https://javarush.com/quests/lectures/questsyntaxpro.level15.lecture00), чтобы не забывать закрывать поток в конце работы

ViktorSalk replied:
Использовал try with resources.

afomkina commented on src/main/java/manager/InMemoryTaskManager.java:
```diff
@@ -240,4 +308,11 @@ public Optional<Epic> getEpicByIdOptional(int id) {
     public Optional<Subtask> getSubtaskByIdOptional(int id) {
```
Нужно удалить все методы, которые возвращают `Optional`, они лишние

ViktorSalk replied:
Удалил все методы, которые возвращают Optional.

afomkina commented on src/main/java/manager/KVServer.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class KVServer {
```
Этот класс нужно использовать

ViktorSalk replied:
Реализовал KVServer и затестил в TaskEndpointTest

afomkina commented on src/main/java/manager/KVServer.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void register(HttpExchange h) throws IOException {
+        try {
```
Нужно использовать try with resources

ViktorSalk replied:
Использовал try with resources, чтобы не забывать закрывать поток в конце работы.

afomkina commented on src/main/java/manager/KVServer.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void save(HttpExchange h) throws IOException {
+        try {
```
Нужно использовать try with resources

ViktorSalk replied:
Использовал try with resources.

afomkina commented on src/main/java/manager/KVServer.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void load(HttpExchange h) throws IOException {
+        try {
```
Нужно использовать try with resources

ViktorSalk replied:
Использовал try with resources.

afomkina commented on src/main/java/manager/KVServer.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private String readText(HttpExchange h) throws IOException {
+        return new String(h.getRequestBody().readAllBytes(), "UTF-8");
```
Лучше использовать встроенную константу `StandardCharsets.UTF_8`

ViktorSalk replied:
Использовал встроенную константу StandardCharsets.UTF_8.

afomkina commented on src/main/java/manager/KVServer.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void sendText(HttpExchange h, String text) throws IOException {
+        byte[] resp = text.getBytes("UTF-8");
```
Лучше использовать встроенную константу `StandardCharsets.UTF_8`

ViktorSalk replied:
Использовал встроенную константу StandardCharsets.UTF_8.

afomkina commented on src/main/java/task/EpicHandler.java:
```diff
@@ -0,0 +1,186 @@
+            }
+        } catch (Exception e) {
+            System.out.println("Произошла ошибка: " + e.getMessage());
+            e.printStackTrace();
```
Печатать стектрейс не нужно

ViktorSalk replied:
Убрал лишний стек-трейсинг.

afomkina commented on src/main/java/task/EpicHandler.java:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    private void handleGet(HttpExchange exchange, String path) throws IOException {
+        try {
```
Используй try with resources здесь и ниже

ViktorSalk replied:
Использовал try with resources в EpicHandler.

afomkina commented on src/main/java/task/PrioritizedHandler.java:
```diff
@@ -0,0 +1,61 @@
+            sendResponse(exchange, gson.toJson(prioritizedTasks), 200); // 200 для GET запроса
+        } catch (Exception e) {
+            System.out.println("Произошла ошибка: " + e.getMessage());
+            e.printStackTrace();
```
Печатать стектрейс не нужно

ViktorSalk replied:
Убрал лишний стек-трейсинг.

afomkina commented on src/main/java/task/SubtaskHandler.java:
```diff
@@ -0,0 +1,124 @@
+            }
+        } catch (Exception e) {
+            System.out.println("Произошла ошибка: " + e.getMessage());
+            e.printStackTrace();
```
Печатать стектрейс не нужно

ViktorSalk replied:
Убрал лишний стек-трейсинг.

afomkina commented on src/main/java/task/TaskHandler.java:
```diff
@@ -0,0 +1,147 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        try {
```
Код со строк 25-47 повторяется в классах `EpicHandler` и `SubtaskHandler`, нужно избавиться от дублирования

ViktorSalk replied:
Избавился от дублирования создав абстрактный класс AbstractTaskHandler и унаследовал handle из него.

Viktor3642 committed:
Оптимизировал API, в котором эндпоинты соответствовуют вызовам базовых методов интерфейса TaskManager.
1) Создал отдельное перечисление для HTTP-кодов и использовать его вместо «голых» цифр в задачх, подзадачах, эпиках, приоритетах.
2) Убрал лишний стек-трейсинг.
3) переименовал однобуквенные переменные в коде.
4) Использовал try with resources, чтобы не забывать закрывать поток в конце работы.
5) Удалил все методы, которые возвращают Optional.
6) Реализовал KVServer и затестил в TaskEndpointTest
7) Использовал встроенную константу StandardCharsets.UTF_8.
8) Избавился от дублирования создав абстрактный класс AbstractTaskHandler и унаследовал дубли из него.

afomkina commented on src/main/java/http/HttpStatusCode.java:
```diff
@@ -0,0 +1,21 @@
+package http;
+
+public enum HttpStatusCode {
```
Клево, что сделал дополнительное задание, молодец 👍

ViktorSalk replied:
Спасибо =)
